File: solutions/0623. Add One Row to Tree/0623.cpp
// Time complexity: O(n)
// Space complexity: O(h)
class Solution {
 public:
  TreeNode* addOneRow(TreeNode* root, int val, int depth) {
    if (depth == 1)
      return new TreeNode(val, root, nullptr);

    if (root == nullptr)
      return new TreeNode(val);    

    addRow(root, val, depth, 1);

    return root;    
  }
 private:
  void addRow(TreeNode* node, int val, int depth, int currDepth) {
    if (node == nullptr)
      return;

    if (currDepth == (depth - 1)) {
      node->left = new TreeNode(val, node->left, nullptr);
      node->right = new TreeNode(val, nullptr, node->right);
      return;
    }

    addRow(node->left, val, depth, currDepth + 1);
    addRow(node->right, val, depth, currDepth + 1);
  }    
};

File: solutions/0623. Add One Row to Tree/0623.py
# Time complexity: O(n)
# Space complexity: O(h)
class Solution:
  def addOneRow(self, root: Optional[TreeNode], 
                val: int, depth: int) -> Optional[TreeNode]:
    if depth == 1:
      return TreeNode(val, root, None)
    
    if root is None:
      return TreeNode(val)
    
    self.addRow(root, val, depth, 1)

    return root
  
  def addRow(self, node: TreeNode, val: int, 
             depth: int, currDepth: int) -> None:
    if node is None:
      return
    
    if currDepth == depth - 1:
      node.left = TreeNode(val, node.left, None)
      node.right = TreeNode(val, None, node.right)
      return
    
    self.addRow(node.left, val, depth, currDepth + 1)
    self.addRow(node.right, val, depth, currDepth + 1)